src/puny.cpp

```diff
@@ -0,0 +1,385 @@
+      if (c < n && (++delta == 0)) { // overflow
+        return std::nullopt;
+      }
+      if (c == n) {
+        for (uint32_t q = delta, k = 36;; k += 36) {
+          uint32_t t = k <= bias ? 1 : (k >= bias + 26 ? 26 : k - bias);
+          if (q < t) {
+            out.push_back(uint8_t(q < 26 ? q + 97 : q + 22));
+            break;
+          }
+          uint32_t char_value = t + (q - t) % (36 - t); // unfortunate division
+          out.push_back(uint8_t(char_value < 26 ? char_value + 97 : char_value + 22));
+
+          q = (q - t) / (36 - t);
+        }
+
+        bias = adapt(delta, processed + 1, basic_count == processed);
+        delta = 0;
+        processed++;
+      }
+    }
+  }
+
+  return out;
+}
+
+// This function merely verifies that a punycode is valid.
+// Important: we do not bother decoding to a buffer.
+bool verify_punycode(std::string_view input) {
+  // So we ignore the first 'xn--'
+  input.remove_prefix(4);
+  if (input.empty()) {
+    return false;
+  }
+  auto loc = input.rfind('-');
+  uint32_t count = 0;
+  if (loc != std::string_view::npos) {
+    for(uint8_t c : input.substr(0, loc)) {
+      if ((c >= 128) || ada::unicode::is_forbidden_domain_code_point(c)) {
+        return false;
+      }
+    }
+    count = uint32_t(loc);
+    input.remove_prefix(loc + 1);
+  }
+  uint32_t n = 128;
+  int i = 0;
+  int bias = 72;
+  for (auto iterator = input.begin(); iterator != input.end();) {
+    int start_i = i;
+    int w = 1;
+    for (int k = 36;; k += 36) {
+      if (iterator == input.end()) {
+        return false;
+      }
+      char code_point = *iterator++;
+      int digit = ((code_point <= 'z') && (code_point >= 'a'))
+                      ? code_point - 'a'
+                      : (((code_point <= '0') && (code_point >= '0'))
+                             ? code_point - '0' + 26
+                             : -1);
+      if (digit < 0) {
+        return false;
+      }
+      i = i + digit * w;
+      int t = k <= bias ? 1 : k >= bias + 26 ? 26 : k - bias;
+      if (digit < t) {
+        break;
+      }
+      w = w * (36 - t);
+    }
+    bias = adapt(i - start_i, count + 1, start_i == 0);
+    n = n + i / (count + 1);
+    i = i % (count + 1);
+    if ((n < 0x80) || is_disallowed(n)) {
+      return false;
+    }
+
+    count++;
+    i++;
+  }
+
+  return true;
+}
+
+// We return the empty string on error.
+std::optional<std::string> convert_domain_to_puny(std::string_view input, bool be_strict) {
+  std::string out;
+  (void)be_strict; // currently ignored.
+  // We seem to be allowing an infinite domain size and an infinite number of
+  // of labels. This is not correct. A domain is limited to 255 characters
+  // and labels cannot exceed 64 characters. However, wpt_tests seems to want
+  // to allow more general cases?
+  // Though wpt_tests does not want limits, let us put one anyhow. If someone
+  // has a domain with over 1MB, we refuse to work on it (safety!).
```

Comment:
Should we add this to readme? Sounds like a good feature.

Reply:
ICU will check for violation of these conditions...

https://github.com/unicode-org/icu/blob/bb0e745e25c99cc57055caf45c81b95ef63b25d4/icu4c/source/common/uts46.cpp#L32

https://github.com/ada-url/ada/blob/2a6208eeab09e95db81a4f7113d48f7e8edf4486/src/parser.cpp#L79

But the tests (WPT) seem to expect that this be ignored... Search for the tests...

> "Label longer than 63 code points",

So it wants to allow longer labels. I am not sure I understand.